File: flutter_baidu_yingyan_trace-1.0.0/lib/src/model/analysis/driving_behavior_option.dart
import 'package:flutter_baidu_yingyan_trace/src/model/base_option.dart';
import 'package:flutter_baidu_yingyan_trace/src/model/coord_type.dart';
import 'package:flutter_baidu_yingyan_trace/src/model/track/query_track_process_option.dart';

/// 驾驶行为分析的请求信息类
class DrivingBehaviourAnalysisOption extends BaseOption {
  /// 需要查询的entity的名称，必填，必须为非空字符串。
  String ?entityName;

  /// 开始时间，必选。
  int ?startTime;

  /// 结束时间，必选。
  int ?endTime;

  /// 固定限速值，可选。
  /// 若设置为非0值，则以设置的数值为阈值，轨迹点速度超过该值则认为是超速；
  /// 若不设置，或设置为0，则根据百度地图道路限速数据计算超速点。
  double ?speedingThreshold;

  /// 自定义轨迹分析时需要的阈值，可选。
  /// 若为null，则使用默认的阈值，若设置，则使用指定的各阈值进行轨迹分析。
  DrivingBehaviorThresholdOption ?thresholdOption;

  /// 纠偏选项，用于控制返回坐标的纠偏处理方式。
  /// 在驾驶行为分析方法中，只有其中的mapMatch和transportMode有效。
  /// 可选。若不设置或为null，则采用默认值。不绑路、交通方式为驾车。
  QueryTrackProcessOption ?processOption;

  /// 返回的坐标类型，可选。
  /// 该字段用于控制返回结果中的坐标类型。可选值为：
  /// CoordType.WGS84: GPS全球卫星定位系统使用的坐标系
  /// CoordType.GCJ02：国测局加密坐标
  /// CoordType.BD09LL：百度经纬度坐标
  /// 该参数仅对国内（包含港、澳、台）轨迹有效，海外区域轨迹均返回 wgs84坐标系
  CoordType ?outputCoordType;

  /// 有参构造
  DrivingBehaviourAnalysisOption(
      {required int tag,
      required int serviceId,
      required this.entityName,
      required this.startTime,
      required this.endTime,
      this.speedingThreshold,
      this.thresholdOption,
      this.processOption,
      this.outputCoordType: CoordType.BD09LL})
      : super(tag: tag, serviceId: serviceId);

  /// map => DrivingBehaviourAnalysisOption
  DrivingBehaviourAnalysisOption.fromMap(Map map)
      : assert(map != null,
            'Construct a DrivingBehaviourAnalysisOption，The parameter map cannot be null !'),
        super.fromMap(map) {
    entityName = map['entityName'];
    startTime = map['startTime'];
    endTime = map['endTime'];
    speedingThreshold = map['speedingThreshold'];
    thresholdOption = map['thresholdOption'] == null
        ? null
        : DrivingBehaviorThresholdOption.fromMap(map['thresholdOption']);
    processOption = map['processOption'] == null
        ? null
        : QueryTrackProcessOption.fromMap(map['processOption']);

    if (map['outputCoordType'] != null) {
      int value = map['outputCoordType'] as int;
      if (value >= 0 && value <= 2) {
        outputCoordType = CoordType.values[value];
      }
    }
  }
  @override
  Map toMap() {
    Map map = super.toMap();
    map['entityName'] = this.entityName;
    map['startTime'] = this.startTime;
    map['endTime'] = this.endTime;
    map['speedingThreshold'] = this.speedingThreshold;
    map['thresholdOption'] = this.thresholdOption?.toMap();
    map['processOption'] = this.processOption?.toMap();
    map['outputCoordType'] = this.outputCoordType?.index;
    return map;
  }
}

/// 轨迹分析时需要的阈值，各阈值均有默认值。
class DrivingBehaviorThresholdOption {
  /// 固定限速值，可选。
  /// 若设置为非0值，则以设置的数值为阈值，轨迹点速度超过该值则认为是超速；
  /// 若不设置，或设置为0，则根据百度地图道路限速数据计算超速点。
  double ?speedingThreshold;

  /// 急加速的水平加速度阈值。
  /// 单位：m^2/s，默认值：1.67，仅支持正数
  double ?harshAccelerationThreshold;

  /// 急减速的水平加速度阈值。
  /// 单位：m^2/s，默认值：-1.67，仅支持负数
  double ?harshBreakingThreshold;

  /// 急转弯的向心加速度阈值。
  /// 单位：m^2/s，默认值：5，仅支持正数
  double ?harshSteeringThreshold;

  /// 有参构造
  DrivingBehaviorThresholdOption(
      {this.speedingThreshold,
      this.harshAccelerationThreshold: 1.67,
      this.harshBreakingThreshold: -1.67,
      this.harshSteeringThreshold: 5.0});

  /// map => DrivingBehaviorThresholdOption
  DrivingBehaviorThresholdOption.fromMap(Map map)
      : assert(map != null,
            'Construct a DrivingBehaviorThresholdOption，The parameter map cannot be null !') {
    speedingThreshold = map['speedingThreshold'];
    harshAccelerationThreshold = map['harshAccelerationThreshold'];
    harshBreakingThreshold = map['harshBreakingThreshold'];
    harshSteeringThreshold = map['harshSteeringThreshold'];
  }

  /// DrivingBehaviorThresholdOption => map
  Map toMap() {
    return {
      'speedingThreshold': this.speedingThreshold,
      'harshAccelerationThreshold': this.harshAccelerationThreshold,
      'harshBreakingThreshold': this.harshBreakingThreshold,
      'harshSteeringThreshold': this.harshSteeringThreshold
    };
  }
}

File: flutter_baidu_yingyan_trace-1.0.0/android/src/main/java/com/baidu/flutter/trace/model/analysis/DrivingBehaviorOption.java
package com.baidu.flutter.trace.model.analysis;

import com.baidu.flutter.trace.model.BaseOption;
import com.baidu.flutter.trace.model.CoordType;
import com.baidu.flutter.trace.model.ProcessOption;
import com.baidu.trace.api.analysis.DrivingBehaviorRequest;

/**
 * 驾驶行为请求
 *
 * @author baidu
 */
public final class DrivingBehaviorOption extends BaseOption {

    /**
     * entity标识
     */
    private String entityName;

    /**
     * 开始时间
     */
    private long startTime;

    /**
     * 结束时间
     */
    private long endTime;

    /**
     * 限速值，用于计算超速，单位为：km/h，取值范围[0,200]，默认值为0
     * <pre>
     * 0：根据百度地图道路限速数据计算超速点；
     * 其他数值：以设置的数值为阈值，轨迹点速度超过该值则认为是超速。
     * </pre>
     *
     * @deprecated
     */
    private double speedingThreshold = 0.0;

    /**
     * 阈值选项
     */
    private ThresholdOption thresholdOption = null;

    /**
     * 纠偏选项，可选项为：
     * <p>
     * <pre>
     * needMapMatch：绑路（false:不需要，true:需要），
     * transportMode：交通方式（driving, riding, walking）
     * </pre>
     */
    private ProcessOption processOption;

    /**
     * 返回结果的坐标类型，默认为百度经纬度
     * <p>
     * <pre>
     * 该字段用于控制返回结果中坐标的类型。可选值为：
     * bd09ll：百度经纬度坐标；
     * gcj02：国测局加密坐标；
     *
     * 注：该字段在国外无效，国外均返回wgs84坐标
     * </pre>
     */
    private int outputCoordType = CoordType.bd09ll.ordinal();

    public String getEntityName() {
        return entityName;
    }

    /**
     * 设置entity标识
     *
     * @param entityName
     */
    public void setEntityName(String entityName) {
        this.entityName = entityName;
    }

    public long getStartTime() {
        return startTime;
    }

    /**
     * 设置开始时间
     *
     * @param startTime
     */
    public void setStartTime(long startTime) {
        this.startTime = startTime;
    }

    public long getEndTime() {
        return endTime;
    }

    /**
     * 设置结束时间
     *
     * @param endTime
     */
    public void setEndTime(long endTime) {
        this.endTime = endTime;
    }

    /**
     * @return
     * @deprecated
     */
    public double getSpeedingThreshold() {
        return speedingThreshold;
    }

    /**
     * 设置限速值，用于计算超速，单位为：km/h，取值范围[0,200]，默认值为0。
     * <p>
     * <pre>
     * 0：根据百度地图道路限速数据计算超速点；
     * 其他数值：以设置的数值为阈值，轨迹点速度超过该值则认为是超速。
     * </pre>
     *
     * @param speedingThreshold
     * @deprecated
     */
    public void setSpeedingThreshold(int speedingThreshold) {
        this.speedingThreshold = speedingThreshold;
    }

    public ThresholdOption getThresholdOption() {
        return thresholdOption;
    }

    /**
     * 设置阈值选项
     *
     * @param thresholdOption
     */
    public void setThresholdOption(ThresholdOption thresholdOption) {
        this.thresholdOption = thresholdOption;
    }

    public ProcessOption getProcessOption() {
        return processOption;
    }

    /**
     * 设置纠偏选项，可选项为：
     * <p>
     * <pre>
     * needMapMatch：绑路（false:不需要，true:需要），
     * transportMode：交通方式（driving, riding, walking）
     * </pre>
     *
     * @param processOption
     */
    public void setProcessOption(ProcessOption processOption) {
        this.processOption = processOption;
    }

    public int getOutputCoordType() {
        return outputCoordType;
    }

    /**
     * 返回结果的坐标类型，默认为百度经纬度
     * <p>
     * <pre>
     * 该字段用于控制返回结果中坐标的类型。可选值为：
     * bd09ll：百度经纬度坐标；
     * gcj02：国测局加密坐标；
     *
     * 注：该字段在国外无效，国外均返回wgs84坐标
     * </pre>
     *
     * @param outputCoordType
     */
    public void setOutputCoordType(int outputCoordType) {
        this.outputCoordType = outputCoordType;
    }

    public DrivingBehaviorOption() {
        super();
    }

    /**
     * @param tag       请求标识
     * @param serviceId 轨迹服务ID
     */
    public DrivingBehaviorOption(int tag, long serviceId) {
        super(tag, serviceId);
    }

    /**
     * @param tag        请求标识
     * @param serviceId  轨迹服务ID
     * @param entityName Entity标识
     */
    public DrivingBehaviorOption(int tag, long serviceId, String entityName) {
        super(tag, serviceId);
        this.entityName = entityName;
    }

    /**
     * @param tag               请求标识
     * @param serviceId         轨迹服务ID
     * @param entityName        Entity标识
     * @param startTime         开始时间
     * @param endTime           结束时间
     * @param speedingThreshold 限速值
     * @param processOption     纠偏选项
     * @param coordTypeOutput   返回结果的坐标类型
     * @deprecated
     */
    public DrivingBehaviorOption(int tag, long serviceId, String entityName, long startTime,
                                 long endTime,
                                 int speedingThreshold, ProcessOption processOption,
                                 int coordTypeOutput) {
        super(tag, serviceId);
        this.entityName = entityName;
        this.startTime = startTime;
        this.endTime = endTime;
        this.speedingThreshold = speedingThreshold;
        this.processOption = processOption;
        this.outputCoordType = coordTypeOutput;
    }

    /**
     * @param tag             请求标识
     * @param serviceId       轨迹服务ID
     * @param entityName      Entity标识
     * @param startTime       开始时间
     * @param endTime         结束时间
     * @param thresholdOption 阈值选项
     * @param processOption   纠偏选项
     * @param coordTypeOutput 返回结果的坐标类型
     */
    public DrivingBehaviorOption(int tag, long serviceId, String entityName, long startTime,
                                 long endTime,
                                 ThresholdOption thresholdOption, ProcessOption processOption,
                                 int coordTypeOutput) {
        super(tag, serviceId);
        this.entityName = entityName;
        this.startTime = startTime;
        this.endTime = endTime;
        this.thresholdOption = thresholdOption;
        this.processOption = processOption;
        this.outputCoordType = coordTypeOutput;
    }

    public DrivingBehaviorRequest toDrivingBehaviorRequest() {
        DrivingBehaviorRequest drivingBehaviorRequest = new DrivingBehaviorRequest();
        drivingBehaviorRequest.setTag(tag);
        drivingBehaviorRequest.setServiceId(serviceId);
        drivingBehaviorRequest.setEntityName(entityName);
        drivingBehaviorRequest.setStartTime(startTime);
        drivingBehaviorRequest.setEndTime(endTime);
        drivingBehaviorRequest
                .setCoordTypeOutput(com.baidu.trace.model.CoordType.values()[outputCoordType]);
        if (thresholdOption != null) {
            drivingBehaviorRequest.setThresholdOption(thresholdOption.toThresholdOption());
        }
        if (processOption != null) {
            drivingBehaviorRequest.setProcessOption(processOption.toProcessOption());
        }
        return drivingBehaviorRequest;
    }

    @Override
    public String toString() {
        return "DrivingBehaviorRequest [tag=" + tag + ", serviceId=" + serviceId + ", entityName="
                + entityName
                + ", startTime=" + startTime + ", endTime=" + endTime + ", thresholdOption="
                + thresholdOption
                + ", processOption=" + processOption + ", coordTypeOutput=" + outputCoordType + "]";
    }

}
